libs/features/commerce/src/placeholder.js

```diff
@@ -0,0 +1,153 @@
+import Service from './service.js';
+
+export const FAILED = 'placeholder-failed';
+export const PENDING = 'placeholder-pending';
+export const RESOLVED = 'placeholder-resolved';
+
+const setImmediate = (callback) => setTimeout(callback, 0);
+
+/**
+ * Stores private bucket of values for each instance of placeholder.
+ * @type {WeakMap<HTMLPlaceholderMethods, {
+ *  error?: Error;
+ *  promises: {
+ *    resolve: (HTMLPlaceholderMethods) => void;
+ *    reject: (error: Error) => void;
+ *  }[];
+ *  state: FAILED | PENDING | RESOLVED | undefined;
+ *  timer?: NodeJS.Timeout;
+ *  version: number;
+ * }>}
+ */
+const buckets = new WeakMap();
+
+/** @type {Commerce.HTMLPlaceholderMixin & Record<string, any>} */
+// @ts-ignore
+export const HTMLPlaceholderMethods = {
+  attributeChangedCallback(name, _, value) {
+    this.log?.debug(`Attribute "${name}":`, value);
+    this.update();
+  },
+
+  connectedCallback() {
+    this.log?.debug('Connected:', this.parentElement);
+    this.update();
+  },
+
+  init() {
+    buckets.set(this, {
+      promises: [],
+      state: undefined,
+      version: 0,
+    });
+  },
+
+  onceSettled() {
+    const { error, promises, state } = buckets.get(this);
+    if (RESOLVED === state) return Promise.resolve(this);
+    if (FAILED === state) return Promise.reject(error);
+    return new Promise((resolve, reject) => {
+      promises.push({ resolve, reject });
+    });
+  },
+
+  toggle() {
+    const bucket = buckets.get(this);
+    [FAILED, PENDING, RESOLVED].forEach((state) => {
+      this.classList.toggle(state, state === bucket.state);
+    });
+  },
+
+  toggleResolved(version) {
+    const bucket = buckets.get(this);
+    // skip obsolete asyncs
+    if (version !== bucket.version) return false;
+    bucket.state = RESOLVED;
+    this.toggle();
+    this.log?.debug('Resolved:', {
+      dataset: { ...this.dataset }, node: this, settings: Service.settings,
+    });
+    // allow calling code to perform sync updates of this element
+    // before notifying observers about state change
+    setImmediate(() => {
+      const promises = bucket.promises;
+      bucket.promises = [];
+      promises.forEach(({ resolve }) => resolve(this));
+      this.dispatchEvent(
+        new CustomEvent(RESOLVED, { bubbles: true }),
+      );
+    });
+    return true;
+  },
+
+  toggleFailed(version, error) {
+    const bucket = buckets.get(this);
+    // skip obsolete asyncs
+    if (version !== bucket.version) return false;
```

Comment:
will this lead having unresolved promises for ever?
When we toggle, all the promises should be cleaned up properly.